include/utils/actions/smooth.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2023 UltiMaker
+// CuraEngine is released under the terms of the AGPLv3 or higher
+
+#ifndef UTILS_VIEWS_SMOOTH_H
+#define UTILS_VIEWS_SMOOTH_H
+
+#include <functional>
+#include <numbers>
+#include <set>
+
+#include <range/v3/action/remove_if.hpp>
+#include <range/v3/functional/bind_back.hpp>
+#include <range/v3/iterator/concepts.hpp>
+#include <range/v3/iterator/operations.hpp>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/view/addressof.hpp>
+#include <range/v3/view/concat.hpp>
+#include <range/v3/view/single.hpp>
+#include <range/v3/view/take.hpp>
+
+#include "utils/types/arachne.h"
+#include "utils/types/generic.h"
+#include "utils/types/geometry.h"
+#include "utils/types/get.h"
+
+namespace cura::actions
+{
+
+struct smooth_fn
+{
+    constexpr auto operator()(const std::integral auto max_resolution, const std::floating_point auto fluid_angle) const
+    {
+        return ranges::make_action_closure(ranges::bind_back(smooth_fn{}, max_resolution, fluid_angle));
+    }
+
+    template<class Rng>
+    requires ranges::forward_range<Rng>&& ranges::sized_range<Rng>&& ranges::erasable_range<Rng, ranges::iterator_t<Rng>, ranges::sentinel_t<Rng>> && (utils::point2d<ranges::range_value_t<Rng>> || utils::junctions<Rng>)
```

Comment:
nitpicky comment, add spaces around binary operators

```suggestion
    requires ranges::forward_range<Rng> && ranges::sized_range<Rng> && ranges::erasable_range<Rng, ranges::iterator_t<Rng>, ranges::sentinel_t<Rng>> && (utils::point2d<ranges::range_value_t<Rng>> || utils::junctions<Rng>)
```